torch/csrc/dynamo/compiled_autograd_py.cpp

```diff
@@ -0,0 +1,526 @@
+    {NULL, NULL, 0, NULL}};
+
+static struct PyModuleDef _module = {
+    PyModuleDef_HEAD_INIT,
+    "torch._C._dynamo.autograd_compiler",
+    "Hooks for compiling autograd",
+    -1,
+    _methods};
+
+static TraceState call_begin_capture(
+    PyObject* self,
+    CacheNode& cache,
+    AutogradCompilerCall& compiler_call,
+    size_t num_outputs) {
+  static PyObject* method_name = PyUnicode_InternFromString("begin_capture");
+  THPObjectPtr pyinput(wrap_variable_list(compiler_call.inputs));
+  THPObjectPtr pysizeinput(cache.wrap_dynamic_inputs());
+  THPObjectPtr pyresult(check(PyObject_CallMethodObjArgs(
+      self, method_name, pyinput.get(), pysizeinput.get(), NULL)));
+
+  PyObject *fake_inputs, *fake_sizes;
+  check(PyArg_ParseTuple(pyresult.get(), "OO", &fake_inputs, &fake_sizes));
+  return TraceState(
+      unwrap_variable_list(fake_inputs),
+      cache.unwrap_dynamic_inputs(fake_sizes),
+      compiler_call.accumulate_grad,
+      num_outputs);
+}
+
+static PyObject* call_end_capture(PyObject* self, const variable_list& inputs) {
+  static PyObject* method_name = PyUnicode_InternFromString("end_capture");
+  THPObjectPtr pyinput(wrap_variable_list(inputs));
+  return check(PyObject_CallMethodOneArg(self, method_name, pyinput.get()));
+}
+
+variable_list compiled_autograd(
+    const std::shared_ptr<Node>& graph_root,
+    GraphTask& graph_task,
+    bool accumulate_grad,
+    const edge_list& output_edges) {
+  TORCH_CHECK(
+      output_edges.empty() || !accumulate_grad,
+      "outputs+accumulate_grad not yet implemented")
+  static std::mutex lock;
+  std::lock_guard<std::mutex> lock_guard(lock);
+  pybind11::gil_scoped_acquire gil;
+  NoGradGuard no_grad; // TODO(jansel): double backward
+
+  std::unordered_map<Node*, int> dependencies =
+      std::move(graph_task.dependencies_);
+  std::vector<std::shared_ptr<Node>> worklist{graph_root};
+  worklist.reserve(dependencies.size());
+  std::unordered_map<Node*, NodeCall> node_inputs;
+  node_inputs.emplace(graph_root.get(), NodeCall(graph_root));
+  std::vector<NodeCall> calls;
+  calls.reserve(dependencies.size() + 8);
+  AutogradCompilerCall compiler_call(accumulate_grad);
+  CacheNode* cache = CacheNode::root();
+
+  for (const auto i : c10::irange(output_edges.size())) {
+    auto inp = node_inputs.find(output_edges[i].function.get());
+    if (inp == node_inputs.end()) {
+      inp = node_inputs
+                .emplace(
+                    output_edges[i].function.get(),
+                    NodeCall(output_edges[i].function))
+                .first;
+    }
+    inp->second.mark_output(output_edges[i].input_nr, i);
+  }
+
+  while (!worklist.empty()) {
+    std::shared_ptr<Node> fn = std::move(worklist.back());
+    worklist.pop_back();
+
+    auto node_input_iter = node_inputs.find(fn.get());
+    TORCH_CHECK(node_input_iter != node_inputs.end());
+    NodeCall& node_call = node_input_iter->second;
+
+    { // update cache and gather args into `compiler_call`
```

Comment:
TODO: Convince myself that you cannot have two graphs with differing structure but the same cache keys, which would cause a spurious cache hit